File: backend/text.py
"""Maltese-aware text normalisation, comparison and diffing.

Speech recognisers routinely drop the Maltese diacritics (ġ ħ ż ċ) and mangle `għ`,
which is silent anyway. So grading has to be tolerant of orthography while still
being able to *show* the learner the correct spelling.
"""

from __future__ import annotations

import difflib
import re
import unicodedata

# Diacritic-insensitive fold. Note `għ` → `` because it is silent; this makes
# "għandi" and a recogniser's "andi" compare equal.
_FOLD = str.maketrans({
    "ġ": "g", "Ġ": "g", "ħ": "h", "Ħ": "h",
    "ż": "z", "Ż": "z", "ċ": "c", "Ċ": "c",
    "à": "a", "è": "e", "ì": "i", "ò": "o", "ù": "u",
    "’": "'", "‘": "'", "`": "'", "´": "'",
})

_PUNCT = re.compile(r"[^\w\s'\-]", re.UNICODE)
_WS = re.compile(r"\s+")


def normalise(s: str) -> str:
    """Display-safe normalisation: unicode NFC, tidy whitespace and apostrophes."""
    s = unicodedata.normalize("NFC", s or "").strip()
    s = s.replace("’", "'").replace("‘", "'")
    return _WS.sub(" ", s)


def fold(s: str) -> str:
    """Aggressive fold used only for comparison, never for display."""
    s = normalise(s).lower().translate(_FOLD)
    s = s.replace("gh", "")          # silent għ, and its de-diacriticised form
    s = _PUNCT.sub(" ", s)
    s = s.replace("-", " ").replace("'", "")
    # The article is frequently dropped or misheard; treat it as optional.
    s = re.sub(r"\b(il|l|ic|id|in|ir|is|it|ix|iz)\b\s*", "", s)
    return _WS.sub(" ", s).strip()


def tokens(s: str) -> list[str]:
    return [t for t in normalise(s).split(" ") if t]


def units(s: str) -> list[tuple[str, str]]:
    """Split into (display, comparison-key) pairs.

    Maltese writes the article and fused prepositions onto the next word with a
    hyphen (`mill-Awstralja`, `id-dar`), and recognisers split them about half the
    time. Splitting at the hyphen — while keeping it attached to the left piece for
    display — makes both spellings tokenise identically, so `mill Awstralja` and
    `mill-Awstralja` compare as equal instead of costing a third of the score.
    """
    out: list[tuple[str, str]] = []
    for word in tokens(s):
        parts = word.split("-")
        for i, part in enumerate(parts):
            if not part:
                continue
            display = part + ("-" if i < len(parts) - 1 else "")
            key = fold(part).replace(" ", "")
            out.append((display, key))
    return out


def similarity(a: str, b: str) -> float:
    """0..1 similarity on folded forms."""
    fa, fb = fold(a), fold(b)
    if not fa and not fb:
        return 1.0
    if not fa or not fb:
        return 0.0
    return difflib.SequenceMatcher(None, fa, fb).ratio()


def word_similarity(a: str, b: str) -> float:
    """Token-level F1 on folded words — more forgiving of word order than raw ratio."""
    ta = [k for _, k in units(a) if k]
    tb = [k for _, k in units(b) if k]
    if not ta or not tb:
        return 0.0
    overlap = 0
    pool = list(tb)
    for t in ta:
        match = next((x for x in pool if difflib.SequenceMatcher(None, t, x).ratio() >= 0.8), None)
        if match:
            pool.remove(match)
            overlap += 1
    precision = overlap / len(ta)
    recall = overlap / len(tb)
    if precision + recall == 0:
        return 0.0
    return 2 * precision * recall / (precision + recall)


def score(said: str, target: str) -> float:
    """Blended score used to grade a spoken/typed attempt against a target."""
    return round(0.45 * similarity(said, target) + 0.55 * word_similarity(said, target), 4)


def diff_words(said: str, target: str) -> list[dict]:
    """Word-level diff for the correction card.

    Returns a list of {op, said, target} where op is equal|sub|del|ins.
    `del` = the learner said something extra, `ins` = they left something out.
    """
    ua, ub = units(said), units(target)
    a, fa = [d for d, _ in ua], [k for _, k in ua]
    b, fb = [d for d, _ in ub], [k for _, k in ub]
    out: list[dict] = []
    matcher = difflib.SequenceMatcher(None, fa, fb)
    for op, i1, i2, j1, j2 in matcher.get_opcodes():
        if op == "equal":
            for k in range(i2 - i1):
                out.append({"op": "equal", "said": a[i1 + k], "target": b[j1 + k]})
        elif op == "replace":
            span = max(i2 - i1, j2 - j1)
            for k in range(span):
                out.append({
                    "op": "sub",
                    "said": a[i1 + k] if i1 + k < i2 else "",
                    "target": b[j1 + k] if j1 + k < j2 else "",
                })
        elif op == "delete":
            for k in range(i1, i2):
                out.append({"op": "del", "said": a[k], "target": ""})
        elif op == "insert":
            for k in range(j1, j2):
                out.append({"op": "ins", "said": "", "target": b[k]})
    return out


# ── Definite-article assimilation, used for hints and for generating drills ──

SUN_LETTERS = set("ċdnrstxzżĊDNRSTXZŻ")
VOWELS = set("aeiouàèìòùAEIOU")


def definite(word: str) -> str:
    """Attach the definite article to a Maltese noun, with assimilation."""
    w = normalise(word)
    if not w:
        return w
    first = w[0]
    if first in VOWELS or first in "hHħĦ" or w[:2].lower() == "għ":
        return f"l-{w}"
    # Romance loans with "s impura" (s + voiceless stop) take a prosthetic i-,
    # which also blocks sun-letter assimilation: skola → l-iskola, not *is-skola.
    # Native clusters do not: żmien → iż-żmien, snien → is-snien.
    if len(w) > 1 and first.lower() == "s" and w[1].lower() in "ptkcq":
        return f"l-i{w}"
    if first in SUN_LETTERS:
        return f"i{first.lower()}-{w}"
    return f"il-{w}"


# ── Language classification ────────────────────────────────────────────────
#
# The tutor contract says reply_mt is Maltese and reply_en is English. A weak model
# breaks that constantly — answering in English, or filling the fields the wrong way
# round. Rather than trusting it, every turn is classified on the way out.

# ħ ġ ż ċ and the digraph għ are near-unique to Maltese orthography.
_MT_CHARS = re.compile(r"[ħĦġĠżŻċĊ]|għ", re.I)

_EN_MARKERS = {
    "the", "and", "you", "your", "what", "who", "where", "when", "how", "why",
    "is", "are", "was", "were", "do", "does", "did", "have", "has", "had",
    "with", "for", "from", "that", "this", "would", "will", "can", "could",
    "want", "hello", "good", "morning", "please", "thanks", "thank", "sorry",
    "mean", "means", "say", "said", "tell", "like", "there", "here", "about",
    "name", "person", "i", "am", "my", "we", "they", "she", "her", "his",
    "of", "to", "it", "be", "been", "at", "on", "by", "or", "but", "if",
    "not", "very", "much", "many", "need", "should", "must", "let", "get",
    "going", "come", "see", "know", "think", "time", "day", "yes", "please",
}

# High-frequency inflected forms the decks store only in a base form, plus the
# pronoun set — without these a short, perfectly good Maltese sentence like
# "Sarah huwa ismek. U inti?" scores as unknown.
_MT_EXTRA = {
    "jien", "jiena", "int", "inti", "hu", "huwa", "hi", "hija", "aħna", "intom",
    "huma", "ismi", "ismek", "ismu", "isimha", "jisimni", "jismek", "jisimhom",
    "tiegħi", "tiegħek", "tiegħu", "tagħha", "tagħna", "tagħkom", "tagħhom",
    "għandi", "għandek", "għandu", "għandha", "għandna", "għandkom", "għandhom",
    "mhux", "mhix", "hemm", "hawn", "issa", "illum", "għada", "mela", "iva",
    "grazzi", "bonġu", "saħħa", "jekk", "jogħġbok", "kif", "fejn", "kemm",
    "liema", "għaliex", "għax", "biex", "imma", "jew", "wkoll", "ukoll",
}

_MT_MARKERS_CACHE: set[str] | None = None


def maltese_markers() -> set[str]:
    """Maltese vocabulary drawn from the app's own decks, so it grows with them."""
    global _MT_MARKERS_CACHE
    if _MT_MARKERS_CACHE is not None:
        return _MT_MARKERS_CACHE
    from . import curriculum

    # Words that are also ordinary English/Italian would fire on the wrong language.
    ambiguous = {"u", "le", "ma", "no", "si", "e", "a", "in", "me", "te", "la", "il"}
    markers: set[str] = set()
    try:
        rows = (curriculum._read_tsv(curriculum.VOCAB_TSV)
                + curriculum._read_tsv(curriculum.PHRASES_TSV))
    except Exception:  # noqa: BLE001 — classification must never break a turn
        rows = []
    for row in rows:
        for word in re.split(r"[\s\-]+", (row.get("mt") or "").lower()):
            word = word.strip(".,!?;:'’\"")
            if len(word) >= 2 and word not in ambiguous:
                markers.add(word)
    markers |= {"mill", "fil", "tal", "bil", "sal", "mal", "għall", "lill", "fl", "bl"}
    markers |= _MT_EXTRA
    _MT_MARKERS_CACHE = markers
    return markers


def _words(s: str) -> set[str]:
    return {w.strip(".,!?;:'’\"()").lower() for w in (s or "").split()}


def looks_maltese(s: str) -> bool:
    w = _words(s)
    hits = len(w & maltese_markers())
    return hits >= 2 or (hits >= 1 and bool(_MT_CHARS.search(s or "")))


def looks_english(s: str) -> bool:
    w = _words(s)
    en = len(w & _EN_MARKERS)
    return en >= 2 and en > len(w & maltese_markers())


_PARENTHETICAL = re.compile(r"\s*[\(\[][^\)\]]*[\)\]]\s*$")


def strip_translation(s: str) -> str:
    """Drop a trailing parenthetical gloss — models like to append `(Xi trid tgħid?)`
    to an English sentence, which belongs in reply_en, not reply_mt."""
    return _PARENTHETICAL.sub("", normalise(s)).strip()


PREP_FUSION = {
    "fi": "fil-", "f'": "fil-", "bi": "bil-", "b'": "bil-", "ta'": "tal-",
    "ta": "tal-", "minn": "mill-", "ma'": "mal-", "għal": "għall-",
    "sa": "sal-", "lil": "lill-",
}

# Maltese prepositions must fuse with a following definite article:
#   minn + il-Belt → mill-Belt ;  fi + l-Awstralja → fl-Awstralja
# Written unfused, they are simply wrong. This is the single most common error an
# English speaker makes, and — unlike most grammar — it is fully mechanical, so it
# can be checked without a model. See `lint_fusion`.
_FUSION_RE = re.compile(
    r"\b(minn|fi|f'|bi|b'|ta'|ta|ma'|għal|sa|lil)\s+(l-|il-|i[ċdnrstxzż]-)",
    re.IGNORECASE,
)

# Before a vowel the article reduces to l-, and some fused forms lose their vowel
# with it: fi + l-Awstralja → fl-Awstralja, not *fil-Awstralja.
_FUSION_BEFORE_VOWEL = {
    "minn": "mill-", "fi": "fl-", "f'": "fl-", "bi": "bl-", "b'": "bl-",
    "ta'": "tal-", "ta": "tal-", "ma'": "mal-", "għal": "għall-",
    "sa": "sal-", "lil": "lill-",
}

# Before an assimilated article (is-, id-, ix-…) the preposition keeps its stem and
# takes the doubled consonant: minn + is-sena → mis-sena, fi + id-dar → fid-dar.
_FUSION_STEM = {
    "minn": "mi", "fi": "fi", "f'": "fi", "bi": "bi", "b'": "bi",
    "ta'": "ta", "ta": "ta", "ma'": "ma", "għal": "għa", "sa": "sa", "lil": "li",
}


def lint_fusion(s: str) -> list[dict]:
    """Find unfused preposition + article sequences.

    Returns [{"found", "should_be", "why"}]. Used as a safety net over tutor output:
    a weaker model will sometimes hand back a "corrected" sentence that still
    contains this error, and shipping that to a learner is worse than not correcting
    at all.
    """
    out: list[dict] = []
    for m in _FUSION_RE.finditer(normalise(s)):
        prep_raw, article = m.group(1), m.group(2)
        prep = prep_raw.lower()
        low = article.lower()
        if low == "l-":
            fused = _FUSION_BEFORE_VOWEL.get(prep, "")
        elif low == "il-":
            fused = PREP_FUSION.get(prep, "")
        else:
            # assimilated article: minn + is- → mis-, fi + id- → fid-
            stem = _FUSION_STEM.get(prep, "")
            fused = f"{stem}{article[1]}-" if stem else ""
        if not fused:
            continue
        rest = m.group(0)[len(prep_raw):].lstrip()
        out.append({
            "found": m.group(0),
            "should_be": fused,
            "why": f"{prep} fuses with the article {rest.rstrip('-')}- to make {fused}",
        })
    return out


def apply_fusion(s: str) -> str:
    """Rewrite unfused sequences into their correct fused forms."""
    def _sub(m: re.Match) -> str:
        fixes = lint_fusion(m.group(0))
        return fixes[0]["should_be"] if fixes else m.group(0)

    return _FUSION_RE.sub(_sub, normalise(s))

File: frontend/text.js
/* Maltese text comparison and phonetic keying, ported from backend/text.py and
   backend/phonetics.py so the app can grade without a server.

   This is a translation, not a reimplementation. Every rule, every threshold and
   the order they are applied in mirror the Python, because the Python is what the
   334 recorded transcripts in tests/fixtures were measured against.
   `tests/test_text_parity.py` runs both over the same strings and compares, for
   the same reason the FSRS port has a parity test: two implementations of one
   algorithm drift silently, and the symptom here would be a learner marked wrong
   for a sentence they said correctly.

   Maltese orthography is close to phonemic, so a small set of rewrites absorbs
   almost all of the variation a recogniser introduces:

       ie → long i          għ → silent
       x  → ʃ               ħ, h → h (h is silent word-finally)
       ż  → z               z → ts
       ċ  → tʃ              ġ → dʒ, j → j (as in yes)
       q  → glottal stop, dropped in the soft key */

/* ── Normalisation and folding ───────────────────────────────────────────── */

const FOLD = {
  'ġ': 'g', 'Ġ': 'g', 'ħ': 'h', 'Ħ': 'h',
  'ż': 'z', 'Ż': 'z', 'ċ': 'c', 'Ċ': 'c',
  'à': 'a', 'è': 'e', 'ì': 'i', 'ò': 'o', 'ù': 'u',
  '’': "'", '‘': "'", '`': "'", '´': "'",
};

// Python's `[^\w\s'\-]` with re.UNICODE keeps letters, digits and underscore.
const PUNCT = /[^\p{L}\p{N}_\s'-]/gu;

export function normalise(s) {
  s = (s || '').normalize('NFC').trim().replace(/[’‘]/g, "'");
  return s.replace(/\s+/g, ' ');
}

export function fold(s) {
  s = normalise(s).toLowerCase().replace(/[ġĠħĦżŻċĊàèìòù’‘`´]/g, (c) => FOLD[c] ?? c);
  s = s.replaceAll('gh', '');            // silent għ, and its de-diacriticised form
  s = s.replace(PUNCT, ' ');
  s = s.replaceAll('-', ' ').replaceAll("'", '');
  // The article is frequently dropped or misheard; treat it as optional.
  s = s.replace(/\b(il|l|ic|id|in|ir|is|it|ix|iz)\b\s*/g, '');
  return s.replace(/\s+/g, ' ').trim();
}

export const tokens = (s) => normalise(s).split(' ').filter(Boolean);

/** Split into [display, comparison-key] pairs, breaking at hyphens.

    Maltese writes the article and fused prepositions onto the next word
    (`mill-Awstralja`, `id-dar`) and recognisers split them about half the time.
    Splitting here makes both spellings tokenise identically. */
export function units(s) {
  const out = [];
  for (const word of tokens(s)) {
    const parts = word.split('-');
    parts.forEach((part, i) => {
      if (!part) return;
      out.push([part + (i < parts.length - 1 ? '-' : ''), fold(part).replaceAll(' ', '')]);
    });
  }
  return out;
}

/* ── difflib.SequenceMatcher.ratio(), faithfully ─────────────────────────────

   Not edit distance. Python's ratio is 2*M/T where M is the total size of the
   matching blocks found by its recursive longest-match algorithm — which is not
   the same number an LCS or a Levenshtein ratio gives, and the thresholds in this
   app (0.86 to accept, 0.8 inside word matching) were tuned against *this*
   function. Substituting a different similarity would silently move every
   decision boundary. */

function matchingBlocks(a, b) {
  const b2j = new Map();
  for (let i = 0; i < b.length; i += 1) {
    const ch = b[i];
    if (!b2j.has(ch)) b2j.set(ch, []);
    b2j.get(ch).push(i);
  }

  function longestMatch(alo, ahi, blo, bhi) {
    let besti = alo, bestj = blo, bestsize = 0;
    let j2len = new Map();
    for (let i = alo; i < ahi; i += 1) {
      const newj2len = new Map();
      for (const j of b2j.get(a[i]) || []) {
        if (j < blo) continue;
        if (j >= bhi) break;
        const k = (j2len.get(j - 1) || 0) + 1;
        newj2len.set(j, k);
        if (k > bestsize) { besti = i - k + 1; bestj = j - k + 1; bestsize = k; }
      }
      j2len = newj2len;
    }
    return [besti, bestj, bestsize];
  }

  const queue = [[0, a.length, 0, b.length]];
  const blocks = [];
  while (queue.length) {
    const [alo, ahi, blo, bhi] = queue.pop();
    const [i, j, k] = longestMatch(alo, ahi, blo, bhi);
    if (!k) continue;
    blocks.push([i, j, k]);
    if (alo < i && blo < j) queue.push([alo, i, blo, j]);
    if (i + k < ahi && j + k < bhi) queue.push([i + k, ahi, j + k, bhi]);
  }
  return blocks;
}

export function ratio(a, b) {
  if (!a.length && !b.length) return 1.0;
  const matches = matchingBlocks(a, b).reduce((n, [, , k]) => n + k, 0);
  const total = a.length + b.length;
  return total ? (2.0 * matches) / total : 1.0;
}

/* ── Scoring ─────────────────────────────────────────────────────────────── */

export function similarity(a, b) {
  const fa = fold(a);
  const fb = fold(b);
  if (!fa && !fb) return 1.0;
  if (!fa || !fb) return 0.0;
  return ratio(fa, fb);
}

/** Token-level F1 on folded words — more forgiving of word order than raw ratio. */
export function wordSimilarity(a, b) {
  const ta = units(a).map(([, k]) => k).filter(Boolean);
  const tb = units(b).map(([, k]) => k).filter(Boolean);
  if (!ta.length || !tb.length) return 0.0;
  const pool = [...tb];
  let overlap = 0;
  for (const t of ta) {
    const idx = pool.findIndex((x) => ratio(t, x) >= 0.8);
    if (idx !== -1) { pool.splice(idx, 1); overlap += 1; }
  }
  const precision = overlap / ta.length;
  const recall = overlap / tb.length;
  if (precision + recall === 0) return 0.0;
  return (2 * precision * recall) / (precision + recall);
}

/* `Math.round(x * 10000) / 10000` is the obvious way to round to four places and
   it disagrees with Python: multiplying by 10000 rounds first, so a value just
   below 0.66875 becomes exactly 6687.5 and then rounds up, where Python — working
   from the exact double — correctly gives 0.6687. The parity test caught it on
   `qagħad id-dar`. `toFixed` rounds the exact value, as Python's round() does. */
const round4 = (x) => Number(x.toFixed(4));

export function score(said, target) {
  return round4(0.45 * similarity(said, target) + 0.55 * wordSimilarity(said, target));
}

/** Word-level diff for the correction card. `del` = extra, `ins` = omitted. */
export function diffWords(said, target) {
  const ua = units(said);
  const ub = units(target);
  const a = ua.map(([d]) => d);
  const fa = ua.map(([, k]) => k);
  const b = ub.map(([d]) => d);
  const fb = ub.map(([, k]) => k);
  const out = [];
  for (const [op, i1, i2, j1, j2] of opcodes(fa, fb)) {
    if (op === 'equal') {
      for (let k = 0; k < i2 - i1; k += 1) {
        out.push({ op: 'equal', said: a[i1 + k], target: b[j1 + k] });
      }
    } else if (op === 'replace') {
      const span = Math.max(i2 - i1, j2 - j1);
      for (let k = 0; k < span; k += 1) {
        out.push({ op: 'sub',
          said: i1 + k < i2 ? a[i1 + k] : '',
          target: j1 + k < j2 ? b[j1 + k] : '' });
      }
    } else if (op === 'delete') {
      for (let k = i1; k < i2; k += 1) out.push({ op: 'del', said: a[k], target: '' });
    } else {
      for (let k = j1; k < j2; k += 1) out.push({ op: 'ins', said: '', target: b[k] });
    }
  }
  return out;
}

function opcodes(a, b) {
  const blocks = matchingBlocks(a, b).sort((x, y) => x[0] - y[0] || x[1] - y[1]);
  blocks.push([a.length, b.length, 0]);
  const out = [];
  let i = 0;
  let j = 0;
  for (const [ai, bj, size] of blocks) {
    if (i < ai && j < bj) out.push(['replace', i, ai, j, bj]);
    else if (i < ai) out.push(['delete', i, ai, j, bj]);
    else if (j < bj) out.push(['insert', i, ai, j, bj]);
    if (size) out.push(['equal', ai, ai + size, bj, bj + size]);
    i = ai + size;
    j = bj + size;
  }
  return out;
}

/* ── Phonetic key ────────────────────────────────────────────────────────── */

// Order matters: digraphs are consumed before the single letters inside them.
const RULES = [
  ['għ', ''], ['gh', ''],
  ['ie', 'I'],
  ['ċ', 'C'], ['ch', 'C'], ['c', 'C'],
  ['ġ', 'J'], ['g', 'J'],
  ['x', 'S'],
  ['ż', 'z'], ['ts', 'z'],
  ['q', 'Q'],
  ['ħ', 'h'],
  ['j', 'y'],
];

export function phoneticKey(s) {
  let t = (s || '').trim().toLowerCase().normalize('NFC')
    .replace(/[’`]/g, "'")
    .replace(/[àèìòù]/g, (c) => FOLD[c]);
  for (const [src, dst] of RULES) t = t.replaceAll(src, dst);
  t = t.replaceAll("'", '');
  t = t.replace(/h\b/g, '');                 // h is silent word-finally
  t = t.replace(/[^a-zA-Z\s]/g, ' ').toLowerCase();
  t = t.replace(/(.)\1+/g, '$1');            // collapse doubles
  return t.replace(/\s+/g, ' ').trim();
}

export const keyNospace = (s) => phoneticKey(s).replaceAll(' ', '');

/** `q` is dropped: recognisers write the glottal stop as għ, as nothing, or as
    whatever consonant follows — never as a consonant of its own. */
export const softKey = (s) => keyNospace(s).replaceAll('q', '');

export function phoneticSimilarity(a, b, soft = false) {
  const ka = soft ? softKey(a) : keyNospace(a);
  const kb = soft ? softKey(b) : keyNospace(b);
  if (!ka && !kb) return 1.0;
  if (!ka || !kb) return 0.0;
  return ratio(ka, kb);
}

/* ── Grading ─────────────────────────────────────────────────────────────── */

export const AGAIN = 1, HARD = 2, GOOD = 3, EASY = 4;

/** Mirrors backend `_assess`: a phonetic floor under the orthographic score, so
    the recogniser's word boundaries do not cost the learner marks. */
export function assess(said, target) {
  const phon = phoneticSimilarity(said, target, true);
  const s = round4(Math.max(phon, 0.6 * phon + 0.4 * score(said, target)));
  return {
    said: normalise(said),
    target: normalise(target),
    score: s,
    grade: s >= 0.95 ? EASY : s >= 0.78 ? GOOD : s >= 0.55 ? HARD : AGAIN,
    verdict: s >= 0.95 ? 'perfect' : s >= 0.75 ? 'close' : s >= 0.5 ? 'partial' : 'off',
    diff: diffWords(said, target),
  };
}
